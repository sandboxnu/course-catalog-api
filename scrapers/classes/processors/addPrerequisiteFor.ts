/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import { BaseProcessor } from "./baseProcessor";
import keys from "../../../utils/keys";
import macros from "../../../utils/macros";
import { isCourseReq, Requisite } from "../../../types/types";
import { ParsedCourseSR, ParsedTermSR } from "../../../types/scraperTypes";

/**
 * Adds the prerequisite-for field for classes that are a predecessor for
 * other classes.
 */
class AddPreRequisiteFor extends BaseProcessor {
  classMap = {};

  /**
   * Creates a class hash map based on the term dump, then calls parse every
   * request to link data.
   *
   * @param {Term Dump} termDump the termDump of the semester.
   */
  go(termDump: ParsedTermSR): void {
    for (const aClass of termDump.classes) {
      const key = keys.getClassHash(aClass);

      // Reset all the prereqsFor arrays at the beginning of each time this is ran over a termDump.
      this.initializeArray(aClass);

      this.classMap[key] = aClass;
    }

    for (const aClass of termDump.classes) {
      if (aClass.prereqs) {
        this.parsePreReqs(aClass, aClass.prereqs, true);
      }
    }

    for (const aClass of termDump.classes) {
      this.sortPreReqs(aClass);
    }
  }

  /**
   * Recursively traverse the prerequisite structure.
   *
   * @param {Class Object} mainClass - the class that we're checking the
   * prereqs for. If it has a prereq, we add this class to the prereq's
   * optPrereqFor field.
   * @param {Requisite} node - a prerequisite class of mainClass. This is
   * the field where we add the mainClass information to.
   * @param {Boolean} isRequired - whether or not the prerequisite is required.
   * @example
   * parsePreReqs({... "subject":"CS","classId":"3500" ...},
   *              {"type":"or","values":[{"classId":"2510","subject":"CS"},
   *                                     {"classId":"1500","subject":"CS","missing":true},
   *                                     {"classId":"2560","subject":"EECE"}],
   *              true})
   */
  parsePreReqs(
    mainClass: ParsedCourseSR,
    node: Requisite,
    isRequired: boolean
  ): void {
    if (typeof node === "string" || (isCourseReq(node) && node.missing)) {
      return;
    }

    // Get the the class we wish to refer to
    if (isCourseReq(node)) {
      const find = keys.getClassHash({
        host: mainClass.host,
        termId: mainClass.termId,
        subject: node.subject,
        classId: node.classId,
      });

      const nodeRef = this.classMap[find];

      if (!nodeRef) {
        macros.error("Unable to find ref for", find, node, mainClass);
        return;
      }

      const classData = {
        subject: mainClass.subject,
        classId: mainClass.classId,
      };

      if (isRequired) {
        nodeRef.prereqsFor.values.unshift(classData);
      } else {
        nodeRef.optPrereqsFor.values.unshift(classData);
      }
    } else {
      const classType = node.type;

      if (node.values !== undefined) {
        node.values.map((course) => {
          // A required course becomes effectively optional when we encounter
          // an 'or' in our tree.
          const reqType = classType === "and" ? isRequired : false;
          return this.parsePreReqs(mainClass, course, reqType);
        });
      }
    }
  }

  /**
   * Creates the fields 'optPrereqsFor' and 'prereqsFor' in nodeRef.
   *
   * @param {Class} nodeRef the class in our tree that we're creating the
   * arrays for.
   */
  initializeArray(nodeRef: ParsedCourseSR): void {
    // Creates the optPrereqsFor field in our class.
    nodeRef.optPrereqsFor = {
      values: [],
    };

    nodeRef.prereqsFor = {
      values: [],
    };
  }

  // Sorts the prereqs for in alphabetical order.
  // except that the classes with the same subject as the main classe's subject.
  // If two classes have the same subject, they are sorted by classId
  sortPrereqsValues(matchingSubject: string, values: Requisite[]): Requisite[] {
    return values.sort((a, b) => {
      if (!(isCourseReq(a) && isCourseReq(b))) {
        return 0;
      }

      if (a.subject !== b.subject) {
        if (a.subject === matchingSubject) {
          return -1;
        }
        if (b.subject === matchingSubject) {
          return 1;
        }
        if (a.subject < b.subject) {
          return -1;
        }
        if (a.subject > b.subject) {
          return 1;
        }
      }

      const firstId = parseInt(a.classId, 10);
      const secondId = parseInt(b.classId, 10);

      if (firstId < secondId) {
        return -1;
      } else if (firstId > secondId) {
        return 1;
      } else {
        return 0;
      }
    });
  }

  /**
   * Recursively traverse the prerequisite structure.
   *
   * @param {Class} aClass - a class to sort the optPrereqsFor and prereqsFor of.
   */
  sortPreReqs(aClass: ParsedCourseSR): void {
    if (aClass.optPrereqsFor && aClass.optPrereqsFor.values) {
      aClass.optPrereqsFor.values = this.sortPrereqsValues(
        aClass.subject,
        aClass.optPrereqsFor.values
      );
    }

    if (aClass.prereqsFor && aClass.prereqsFor.values) {
      aClass.prereqsFor.values = this.sortPrereqsValues(
        aClass.subject,
        aClass.prereqsFor.values
      );
    }
  }
}

export default new AddPreRequisiteFor();
