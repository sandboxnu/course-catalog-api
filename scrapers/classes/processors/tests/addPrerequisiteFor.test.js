/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */
import addPrerequisiteFor from "../addPrerequisiteFor";

describe("addPrerequisiteForFor tests", () => {
  const cs2500 = {
    prereqs: {
      type: "or",
      values: [
        {
          subject: "CS",
          classId: "3000",
        },
        "this is a string2",
      ],
    },
    classId: "2500",
    termId: "201829",
    subject: "CS",
    host: "neu.edu",
  };

  const cs3000 = {
    classId: "3000",
    termId: "201829",
    subject: "CS",
    host: "neu.edu",
  };

  const engw3302 = {
    prereqs: {
      type: "and",
      values: [
        {
          subject: "CS",
          classId: "2500",
        },
        {
          subject: "not in term dump",
          classId: "1234",
        },
        "this is a string",
        {
          type: "or",
          values: [
            {
              subject: "CS",
              classId: "3000",
            },
            "this is a string2",
          ],
        },
      ],
    },
    classId: "3302",
    termId: "201829",
    subject: "ENGW",
    host: "neu.edu",
  };

  const cs1234 = {
    prereqs: {
      type: "or",
      values: [
        {
          subject: "CS",
          classId: "3000",
        },
        {
          type: "or",
          values: undefined,
        },
      ],
    },
    classId: "1234",
    termId: "201829",
    subject: "CS",
    host: "neu.edu",
  };

  const termDump = {
    classes: [cs2500, cs3000, engw3302, cs1234],
    sections: [],
  };

  const cs2500Parsed = {
    ...JSON.parse(JSON.stringify(cs2500)),
    optPrereqsFor: {
      values: [],
    },
    prereqsFor: {
      values: [
        {
          classId: "3302",
          subject: "ENGW",
        },
      ],
    },
  };

  const cs3000Parsed = {
    ...JSON.parse(JSON.stringify(cs3000)),
    optPrereqsFor: {
      values: [
        {
          classId: "1234",
          subject: "CS",
        },
        {
          classId: "2500",
          subject: "CS",
        },
        {
          classId: "3302",
          subject: "ENGW",
        },
      ],
    },
    prereqsFor: {
      values: [],
    },
  };

  const engw3302Parsed = {
    ...JSON.parse(JSON.stringify(engw3302)),
    optPrereqsFor: {
      values: [],
    },
    prereqsFor: {
      values: [],
    },
  };

  const cs1234Parsed = {
    ...JSON.parse(JSON.stringify(cs1234)),
    optPrereqsFor: {
      values: [],
    },
    prereqsFor: {
      values: [],
    },
  };

  it("should parse prereqs", () => {
    addPrerequisiteFor.go(termDump);
    expect(termDump).toEqual({
      classes: [cs2500Parsed, cs3000Parsed, engw3302Parsed, cs1234Parsed],
      sections: [],
    });
  });

  it("should sort some optPrereqsFor", () => {
    const cs3000Sorted = {
      ...JSON.parse(JSON.stringify(cs3000)),
      optPrereqsFor: {
        values: [
          {
            classId: "3302",
            subject: "ENGW",
          },
          {
            classId: "2500",
            subject: "CS",
          },
          {
            classId: "1234",
            subject: "CS",
          },
        ],
      },
      prereqsFor: {
        values: [],
      },
    };
    addPrerequisiteFor.sortPrereqs(cs3000Sorted);
    expect(cs3000Sorted.optPrereqsFor).toEqual(cs3000Parsed.optPrereqsFor);
  });

  it("should sort some prereqsFor", () => {
    const cs2500Sorted = {
      ...JSON.parse(JSON.stringify(cs2500)),
      optPrereqsFor: {
        values: [],
      },
      prereqsFor: {
        values: [
          "string",
          {
            classId: "3302",
            subject: "CS",
          },
          {
            classId: "3000",
            subject: "ENGW",
          },
          {
            classId: "3302",
            subject: "ENGW",
          },
          {
            classId: "3302",
            subject: "MATH",
          },
        ],
      },
    };

    addPrerequisiteFor.sortPrereqs(cs2500Sorted);
    expect(cs2500Sorted.prereqsFor).toEqual({
      values: [
        "string",
        {
          classId: "3302",
          subject: "CS",
        },
        {
          classId: "3000",
          subject: "ENGW",
        },
        {
          classId: "3302",
          subject: "ENGW",
        },
        {
          classId: "3302",
          subject: "MATH",
        },
      ],
    });
  });
});
