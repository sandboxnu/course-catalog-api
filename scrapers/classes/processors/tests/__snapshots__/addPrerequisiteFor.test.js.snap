// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addPreRequisiteFor tests should sort some optPrereqsFor 1`] = `
Object {
  "values": Array [
    Object {
      "classId": "5",
      "subject": "CS",
    },
    Object {
      "classId": "7",
      "subject": "EECE",
    },
    Object {
      "classId": "11",
      "subject": "EECE",
    },
    Object {
      "classId": "2",
      "subject": "MATH",
    },
    Object {
      "classId": "3",
      "subject": "MATH",
    },
  ],
}
`;

exports[`addPreRequisiteFor tests should sort some prereqsFor 1`] = `
Object {
  "values": Array [
    Object {
      "classId": "7",
      "subject": "EECE",
    },
    Object {
      "classId": "11",
      "subject": "EECE",
    },
    Object {
      "classId": "5",
      "subject": "CS",
    },
    Object {
      "classId": "2",
      "subject": "MATH",
    },
    Object {
      "classId": "3",
      "subject": "MATH",
    },
  ],
}
`;
